sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -501,6 +502,46 @@ object RemoveNoopOperators extends Rule[LogicalPlan] {
   }
 }
 
+/**
+ * Remove no-op `Union` from the query plan that do not make any modifications.
+ */
+object RemoveNoopUnion extends Rule[LogicalPlan] {
+  /**
+   * This only removes the `Project` that has only attributes or aliased attributes
+   * from its child.
+   */
+  private def removeAliasOnlyProject(plan: LogicalPlan): LogicalPlan = plan match {
+    case p @ Project(projectList, child) =>
```

Comment:
How about a `Aggregate` case?

Reply:
You mean a no-op Aggregation like `select key as key2, value from table group by key value`?

Reply:
I thought it was like this:
```
scala> sql("select distinct * from (select id from t1 group by id union select id as key from t2 group by id)").explain(true)

== Analyzed Logical Plan ==
id: bigint
Distinct
+- Project [id#6L]
   +- SubqueryAlias __auto_generated_subquery_name
      +- Distinct
         +- Union false, false
            :- Aggregate [id#6L], [id#6L]
            :  +- SubqueryAlias spark_catalog.default.t1
            :     +- Relation[id#6L] parquet
            +- Aggregate [id#7L], [id#7L AS key#13L]
               +- SubqueryAlias spark_catalog.default.t2
                  +- Relation[id#7L] parquet

== Optimized Logical Plan ==
Aggregate [id#6L], [id#6L]
+- Aggregate [id#6L], [id#6L]
   +- Union false, false
      :- Aggregate [id#6L], [id#6L]
      :  +- Relation[id#6L] parquet
      +- Aggregate [id#7L], [id#7L AS key#13L]
         +- Relation[id#7L] parquet
```

Reply:
I thought it was like this:
```
scala> sql("select distinct * from (select id from t group by id union select id as key from t group by id)").explain(true)
...
== Analyzed Logical Plan ==
id: bigint
Distinct
+- Project [id#8L]
   +- SubqueryAlias __auto_generated_subquery_name
      +- Distinct
         +- Union false, false
            :- Aggregate [id#8L], [id#8L]
            :  +- SubqueryAlias spark_catalog.default.t
            :     +- Relation[id#8L] parquet
            +- Aggregate [id#8L], [id#8L AS key#7L]
               +- SubqueryAlias spark_catalog.default.t
                  +- Relation[id#8L] parquet

== Optimized Logical Plan ==
Aggregate [id#8L], [id#8L]
+- Aggregate [id#8L], [id#8L]
   +- Union false, false
      :- Aggregate [id#8L], [id#8L]
      :  +- Relation[id#8L] parquet
      +- Aggregate [id#8L], [id#8L AS key#7L]
         +- Relation[id#8L] parquet
```

Reply:
yea, that is what I said a no-op aggregate. I think it is okay to add it. Maybe I can add it in follow up.

Reply:
yea, sgtm.